File: src/contracts/note.ts
export type NoteScope = "active" | "archived" | "deleted";
export type NoteColor = "lemon" | "peach" | "rose" | "lilac" | "sky" | "mint";
/** Built-in paper or a user color added from the settings color pool. */
export type NoteColorId = NoteColor | (string & {});
export type TextDirection = "automatic" | "ltr" | "rtl";

/** One swatch of the note color pool. User colors carry their own hex `value`. */
export interface NoteColorOption {
  id: string;
  name: string;
  value?: string;
}

export const CUSTOM_NOTE_COLOR_PREFIX = "custom-";
export const MAX_CUSTOM_NOTE_COLORS = 12;

export interface NoteRecord {
  id: string;
  title: string;
  body: string;
  color: NoteColorId;
  createdAtMs: number;
  updatedAtMs: number;
  archivedAtMs: number | null;
  deletedAtMs: number | null;
  sortKey: string;
  textDirection: TextDirection;
  revision: number;
}

export interface ListNotesQuery {
  scope: NoteScope;
  query: string;
}

export interface CreateNoteInput {
  title: string;
  body: string;
  color: NoteColorId;
  textDirection: TextDirection;
}

export interface UpdateNoteInput extends CreateNoteInput {
  id: string;
  expectedRevision: number;
}

export interface NoteMutationInput {
  id: string;
  expectedRevision: number;
}

File: src-tauri/src/domain/note.rs
use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum NoteScope {
    Active,
    Archived,
    Deleted,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListNotesQuery {
    pub scope: NoteScope,
    #[serde(default)]
    pub query: String,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateNoteInput {
    pub title: String,
    pub body: String,
    pub color: String,
    #[serde(default = "default_text_direction")]
    pub text_direction: String,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateNoteInput {
    pub id: String,
    pub title: String,
    pub body: String,
    pub color: String,
    pub text_direction: String,
    pub expected_revision: i64,
}

#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NoteMutationInput {
    pub id: String,
    pub expected_revision: i64,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct NoteRecord {
    pub id: String,
    pub title: String,
    pub body: String,
    pub color: String,
    pub created_at_ms: i64,
    pub updated_at_ms: i64,
    pub archived_at_ms: Option<i64>,
    pub deleted_at_ms: Option<i64>,
    pub sort_key: String,
    pub text_direction: String,
    pub revision: i64,
}

pub fn default_text_direction() -> String {
    "automatic".into()
}
